presto-main/src/test/java/com/facebook/presto/block/TestBlockBuilder.java

```diff
@@ -108,6 +112,24 @@ public void testNewBlockBuilderLike()
         assertEquals(newArrayBlockBuilder.getPositionCount(), 1);
         // Reserved 200 ARRAY(ARRAY(BIGINT)), and 201 ints for offsets and 200 booleans for nulls
         assertEquals(newArrayBlockBuilder.getRetainedSizeInBytes(), 5848);
+
+        //------------------------- Test newBlockBuilderLike(BlockBuilderStatus, int) on large BlockBuilders so that there is no Integer overflow-------------------------
```

Comment:
The need to add these separators suggests that this test needs to be split into multiple.

Reply:
Sure. I created a new test `testNewBlockBuilderLikeWithExpectedEntries` for the last two sections separated by the separators.
```
@Test
    public void testNewBlockBuilderLikeWithExpectedEntries()
    {
        ArrayType longArrayType = new ArrayType(BIGINT);
        ArrayType arrayType = new ArrayType(longArrayType);
        List<Type> channels = ImmutableList.of(BIGINT, VARCHAR, arrayType);
        PageBuilder pageBuilder = new PageBuilder(channels);
        BlockBuilder bigintBlockBuilder = pageBuilder.getBlockBuilder(0);
        BlockBuilder varcharBlockBuilder = pageBuilder.getBlockBuilder(1);
        BlockBuilder arrayBlockBuilder = pageBuilder.getBlockBuilder(2);

        BlockBuilder newBigintBlockBuilder = bigintBlockBuilder.newBlockBuilderLike(null, 200);
        assertEquals(newBigintBlockBuilder.getPositionCount(), 0);
        assertEquals(newBigintBlockBuilder.getRetainedSizeInBytes(), 80);
        newBigintBlockBuilder.writeLong(0);
        assertEquals(newBigintBlockBuilder.getPositionCount(), 1);
        // Reserved 200 longs and booleans for nulls array
        assertEquals(newBigintBlockBuilder.getRetainedSizeInBytes(), 1880);

        BlockBuilder newVarcharBlockBuilder = varcharBlockBuilder.newBlockBuilderLike(null, 200);
        assertEquals(newVarcharBlockBuilder.getPositionCount(), 0);
        assertEquals(newVarcharBlockBuilder.getRetainedSizeInBytes(), 164);
        newVarcharBlockBuilder.writeLong(0);
        newVarcharBlockBuilder.closeEntry();
        assertEquals(newVarcharBlockBuilder.getPositionCount(), 1);
        // Reserved 200 varchars of average length 5.9, and 201 ints for offsets and 200 booleans for nulls
        assertEquals(newVarcharBlockBuilder.getRetainedSizeInBytes(), 2360);

        BlockBuilder newArrayBlockBuilder = arrayBlockBuilder.newBlockBuilderLike(null, 200);
        assertEquals(newArrayBlockBuilder.getPositionCount(), 0);
        assertEquals(newArrayBlockBuilder.getRetainedSizeInBytes(), 248);
        newArrayBlockBuilder.appendStructure(fromElementBlock(1, Optional.empty(), IntStream.range(0, 2).toArray(), newBigintBlockBuilder.build()));
        assertEquals(newArrayBlockBuilder.getPositionCount(), 1);
        // Reserved 200 ARRAY(ARRAY(BIGINT)), and 201 ints for offsets and 200 booleans for nulls
        assertEquals(newArrayBlockBuilder.getRetainedSizeInBytes(), 5848);

        // Test large nested BlockBuilders so that there is no Integer overflow

        pageBuilder.reset();
        BlockBuilder largeVarcharBlockBuilder = pageBuilder.getBlockBuilder(1);

        // Construct a string of 64 * 16 = 2^11 bytes
        Slice largeSlice = Slices.utf8Slice(String.join("", Collections.nCopies(64, "CowMonsterKing:)")));
        // Write the string to the largeVarcharBlockBuilder for 2^20 times
        for (int i = 0; i < 1_048_576; i++) {
            VARCHAR.writeSlice(largeVarcharBlockBuilder, largeSlice);
            pageBuilder.declarePosition();
        }
        newVarcharBlockBuilder = largeVarcharBlockBuilder.newBlockBuilderLike(null, 1_048_576 * 8);
        assertEquals(newVarcharBlockBuilder.getPositionCount(), 0);
        assertEquals(newVarcharBlockBuilder.getRetainedSizeInBytes(), 164);

        // We are not going to test real reservation here because allocating large amount of memory fails the Travis.
    }
```

Reply:
Masha, I just noticed the tests on the new newBlockBuilderLike() with expectedEntries need to be called on the previously existing BlockBuilders. So I put the tests back and created a new one
`testNewBlockBuilderLikeForLargeBlockBuilder` just for testing the integer overflow. 
```
@Test
    public void testNewBlockBuilderLikeForLargeBlockBuilder()
    {
        List<Type> channels = ImmutableList.of(VARCHAR);
        PageBuilder pageBuilder = new PageBuilder(channels);
        BlockBuilder largeVarcharBlockBuilder = pageBuilder.getBlockBuilder(0);

        // Construct a string of 64 * 16 = 2^11 bytes
        Slice largeSlice = Slices.utf8Slice(String.join("", Collections.nCopies(64, "CowMonsterKing:)")));
        // Write the string to the largeVarcharBlockBuilder for 2^20 times
        for (int i = 0; i < 1_048_576; i++) {
            VARCHAR.writeSlice(largeVarcharBlockBuilder, largeSlice);
            pageBuilder.declarePosition();
        }
        BlockBuilder newVarcharBlockBuilder = largeVarcharBlockBuilder.newBlockBuilderLike(null, 1_048_576 * 8);
        assertEquals(newVarcharBlockBuilder.getPositionCount(), 0);
        assertEquals(newVarcharBlockBuilder.getRetainedSizeInBytes(), 164);

        // We are not going to test real reservation here because allocating large amount of memory fails the Travis.
    }
```